gramineproject/gramine
# [LibOS] Add support for `flock` system call

## Description of the changes 
We implemented it with the OCALL mechanism to make it compatible with NFS.
```
int flock(int fd, int operation);
```
The function applies or removes an advisory lock on the open file specified by `fd`.
In Linux, We can use `flock` to place an advisory lock in file `fd`, socket `fd` , pipe `fd`, and event `fd`.



## How to test this PR? 
To test the function we should notice the following situation

1. A call to flock() may block if another process holds an incompatible lock. To make a nonblocking request, include LOCK_NB (by ORing) with any of the above operations.
2. A single file may not simultaneously have both shared and exclusive locks.
3. Locks created by flock() are associated with an open file table entry. This means that duplicate file descriptors (created by, for example, [fork](https://linux.die.net/man/2/fork)(2) or [dup](https://linux.die.net/man/2/dup)(2)) refer to the same lock, and this lock may be modified or released using any of these descriptors. Furthermore, the lock is released either by an explicit LOCK_UN operation on any of these duplicate descriptors, or when all such descriptors have been closed.
4. If a process uses [open](https://linux.die.net/man/2/open)(2) (or similar) to obtain more than one descriptor for the same file, these descriptors are treated independently by flock(). An attempt to lock the file using one of these file descriptors may be denied by a lock that the calling process has already placed via another descriptor.

Below is the test code.
```
#include <stdio.h>
#include <unistd.h>
#include <sys/file.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <errno.h>

static struct tcase {
	int operation;
	char *opt;
} tcases[] = {
	{LOCK_SH, "Shared Lock" },
	{LOCK_UN, "Unlock"},
	{LOCK_EX, "Exclusive Lock"},
};

int open_file(int perm) {
    int fd = open("test_flock.txt",perm);
    if (fd == -1)
        perror("open error!\n");

    return fd;
}

static void verify_flock_1(int index, int count) {
    int fd = open_file(O_RDONLY);

    for (size_t i =0; i < sizeof(tcases)/sizeof(tcases[0]);i++) {
        if(flock(fd,tcases[i].operation)) {
            printf("flock() with %s error\n", tcases[i].opt);
            printf("[%d/%d verify_flock_1 FAILED!]\n", index, count);
            return;
        }
    }

    close(fd);
    printf("[%d/%d verify_flock_1 SUCCESS!]\n", index, count);
}

static int badfd = -1;
static int fd;
static struct tcase_2 {
	int *fd;
	int operation;
	int exp_err;
} tcases_2[] = {
	{&badfd, LOCK_SH, EBADF},
	{&fd, LOCK_NB, EINVAL},
	{&fd, LOCK_SH | LOCK_EX, EINVAL},
};

static void verify_flock_2(int index, int count) {
    int tmpfd = open_file(O_RDONLY);
    tcases_2[1].fd = &tmpfd;
    tcases_2[2].fd = &tmpfd;

    for (size_t i =0; i < sizeof(tcases_2)/sizeof(tcases[0]);i++) {
        if(flock(*tcases_2[i].fd, tcases_2[i].operation) == 0) {
            printf("flock() succeeded unexpectedly\n");
            printf("FAILED PARAMETERS: [fd:%d] [operation:%d] [exp_err:%d]\n",*tcases_2[i].fd, tcases_2[i].operation, tcases_2[i].exp_err);
            printf("[%d/%d verify_flock_2 FAILED!]\n", index, count);
            return;
        } else {
            if (errno != tcases_2[i].exp_err) {
                printf("flock() failed unexpectedly\n");
                printf("EXPECTED errno %d\n",tcases_2[i].exp_err);
                return;
            }
        }
    }
    close(tmpfd);
    printf("[%d/%d verify_flock_2 SUCCESS!]\n", index, count);
}

static void childfunc(int fd, int index, int count) {
    sleep(1);
    int fd2 = open_file(O_RDONLY);
    if (flock(fd2, LOCK_EX | LOCK_NB) != -1) {
        printf("CHILD: The file was not already locked!\n");
        goto out;
    }


    if (flock(fd, LOCK_UN) == -1) {
        printf("CHILD: Unable to unlock file locked by parent!\n");
        goto out;
    }

    if (flock(fd2, LOCK_EX | LOCK_NB) == -1) {
        printf("CHILD: Unable to lock file after unlock!\n");
        goto out;
    }

    close(fd2);
    printf("[%d/%d verify_flock_3 SUCCESS!]\n", index, count);
    return;

out:
    printf("[%d/%d verify_flock_3 FAILED!]\n", index, count);
    return;
}

static void verify_flock_3(int index, int count) {
    int fd1;
    pid_t pid;

    fd1 =  open_file(O_RDONLY);

    pid = fork();
    if (pid == 0)
        childfunc(fd1, index, count);
    else {
        if (flock(fd1, LOCK_EX | LOCK_NB) == -1) {
            printf("Parent: Initial attempt to flock() failed\n");
        }
        int status;
        wait(&status);
        close(fd1);
    }
}

static void verify_flock_4(int index, int count) {
    int fd1, fd2;

    fd1 = open_file(O_RDONLY);
    if (flock(fd1, LOCK_EX | LOCK_NB)) {
        printf("FAILED: First attempt to flock() failed\n");
    }

    fd2 = open_file(O_RDONLY);
    if (flock(fd2, LOCK_EX | LOCK_NB)) {
        printf("SUCCESS: Second attempt to flock() denied\n");
    } else {
        printf("FAILED: Second attempt to flock() SUCCESS UNEXPECTLY\n");
    }

    if (flock(fd1, LOCK_UN)) {
        printf("FAILED: Failed to unlock fd1\n");
    }

    if (flock(fd2,  LOCK_EX | LOCK_NB)) {
        printf("FAILED: Third attempt to flock() denied!\n");
    }

    printf("[%d/%d verify_flock_4 SUCCESS!]\n", index, count);
}

int main(void) {
    verify_flock_1(1,4);
    verify_flock_2(2,4);
    verify_flock_3(3,4);
    verify_flock_4(4,4);
}
```


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/1171)
